Main.hs

```diff
@@ -273,30 +276,66 @@ compileConstructorArgs (ExtendTel a tel)
     (:) <$> compileType (unEl $ unDom a) <*> compileConstructorArgs tel
 compileConstructorArgs tel = genericDocError =<< text "Bad constructor args:" <?> prettyTCM tel
 
-compileFun :: Definition -> TCM [Hs.Decl ()]
-compileFun def = do
+compilePostulate :: Definition -> TCM [Hs.Decl ()]
+compilePostulate def = do
   let x = hsName $ prettyShow $ qnameName $ defName def
   ty <- compileType (unEl $ defType def)
-  cs <- mapM (compileClause x) funClauses
+  let body = hsError $ "postulate: " ++ pp ty
+  return [ Hs.TypeSig () [x] ty
+         , Hs.FunBind () [Hs.Match () x [] (Hs.UnGuardedRhs () body) Nothing] ]
+  where
+    Axiom = theDef def
+
+type LocalDecls = [(QName, Definition)]
+
+groupDecls :: LocalDecls -> [(Definition, LocalDecls)]
+groupDecls [] = []
+groupDecls ((q, d) : defs) =
+  let getLvl (QName (MName xs) _) = length xs
+      (children, defs') = span ((> getLvl q) . getLvl . fst) defs
+  in (d, children) : groupDecls defs'
+
+compileFun :: Definition -> TCM [Hs.Decl ()]
+compileFun d = do
+  let findLocals = fromMaybe [] . fmap snd . find ((defName d ==) . defName . fst)
+  locals <- findLocals . groupDecls . sortDefs <$> curDefs
+  compileFun' locals d
+
+compileFun' :: LocalDecls -> Definition -> TCM [Hs.Decl ()]
+compileFun' locals def@(Defn {..}) = do
+  let x = hsName $ prettyShow $ qnameName defName
+  ty <- compileType (unEl defType)
+  cs <- mapM (compileClause locals defName x) funClauses
   return [Hs.TypeSig () [x] ty, Hs.FunBind () cs]
   where
-    Function{..} = theDef def
+    Function{..} = theDef
 
-compileClause :: Hs.Name () -> Clause -> TCM (Hs.Match ())
-compileClause x Clause{clauseTel       = tel,
-                       namedClausePats = ps,
-                       clauseBody      = body } =
+compileClause :: LocalDecls -> QName -> Hs.Name () -> Clause -> TCM (Hs.Match ())
+compileClause locals qn x c@Clause{clauseTel = tel, namedClausePats = ps', clauseBody = body} =
   addContext (KeepNames tel) $ localScope $ do
-    -- Only bind user-written variables
+    -- Compile patterns, only bind user-written variables
     let bind (d, i) | getOrigin d == UserWritten = bindVar i
                     | otherwise                  = return ()
     mapM_ bind $ zip (telToList tel) $ reverse [0..size tel - 1]
-    ps   <- compilePats ps
-    body <- maybe (return $ Hs.Var () $ Hs.UnQual () (hsName "undefined")) compileTerm body
-    let rhs = Hs.UnGuardedRhs () body
+    ps <- compilePats ps'
+
+    -- Compile rhs and its @where@ clauses, making sure that:
+    --   * inner modules get instantiated
+    --   * references to inner modules get un-qualifiyed (and instantiated)
+    let elims = patternsToElims ps'
+        children = groupDecls $ map (fmap (`applyE` elims)) locals
```

Comment:
This is not correct. Local functions are parameterised over the variables bound in the lhs, not over the arguments to the parent function. You need `teleArgs tel` (and `apply` instead of `applyE`).